### PR TITLE
Set device specific prefix cache key.

### PR DESCRIPTION
**Pull request by hans2103**
problem: Joomla! makes no difference in devices when caching a page. 

This pull request makes an end to it and adds a prefix to the cache ID. 
This pull request uses the Joomla! default browser detection implementation.

### CODE REVIEW

**Comment by mbabker** [comment]
2 concerns here:

1) Method name.  setDevicePrefix() implies I can set the prefix myself, in reality you're getting a prefix here.
2) To me it seems like this change really only affects sites which have a massively different mobile version compared to the desktop version.  On sites using fully responsive templates and presenting the same output to all devices, there is no benefit to be gained with this change.

**Comment by hans2103** [comment]
Hi @mbabker 
1. please feel free to suggest a better name. Perhaps another name suits better. 
2. example: with this PR you have the ability to create device specific design. For instance... don't show huge background images in masthead on mobile while showing them on desktop. (The current device detection implementation of Joomla doesn't trigger tablet, bot or other devices). Even a slight difference is good enough for me to cache that different page.

**Comment by zero-24** [comment]
>1. please feel free to suggest a better name. Perhaps another name suits better. 

maybe `getDevicePrefix()`?

File: libraries/joomla/cache/cache.php [hunk 1]
```diff
@@ -748,7 +749,26 @@ public static function makeId()
...
+	{
+		jimport('joomla.environment.browser');
+		$browser = JBrowser::getInstance();
+
+		if($browser->isMobile())
+		{
+			return 'M-';
+		}
+
+		return ;
```
Comment (zero-24) [inline libraries/joomla/cache/cache.php]: can you remove the space between `retun` and  `;`

File: libraries/joomla/cache/cache.php [hunk 2]
```diff
@@ -748,7 +749,26 @@ public static function makeId()
...
+	 *
+	 * @return  string   Device specific prefix
+	 *
+	 */
+	public function setDevicePrefix()
+	{
+		jimport('joomla.environment.browser');
+		$browser = JBrowser::getInstance();
+
+		if($browser->isMobile())
```
Comment (zero-24) [inline libraries/joomla/cache/cache.php]: can you add a space after the `if`

File: libraries/joomla/cache/cache.php [hunk 3]
```diff
@@ -748,7 +749,26 @@ public static function makeId()
...
 
-		return md5(serialize($safeuriaddon));
+		return $prefix . md5(serialize($safeuriaddon));
+	}
+
+	/**
+	 * Set prefix cache key if device calls for separate caching
+	 *
+	 * @return  string   Device specific prefix
+	 *
```
Comment (zero-24) [inline libraries/joomla/cache/cache.php]: can you add a `@since 3.5` here?

Comment (hans2103) [inline libraries/joomla/cache/cache.php]: done

File: libraries/joomla/cache/cache.php (hunk 2 above)
Comment (hans2103) [inline libraries/joomla/cache/cache.php]: done

File: libraries/joomla/cache/cache.php (hunk 1 above)
Comment (hans2103) [inline libraries/joomla/cache/cache.php]: done

**Comment by hans2103** [comment]
Now creating an option for Global Config to enable device specific caching if Caching is activated. Will add this to this pull request.

**Comment by dgt41** [comment]
@hans2103 shouldn’t we first work on #6859 and then try to fix the caching? For me this looks like back and forth...

**Comment by hans2103** [comment]
yes and no... This pull request works with the current implementation of the mobile detect class.
When #6859 is implemented this PR can be expanded with the newly created class.

**Comment by Fedik** [comment]
>problem: Joomla! makes no difference in devices when caching a page.

to be honest I do not see the problem here, and have a doubt that Joomla realy need need it ...
any real life example?

**Comment by hans2103** [comment]
@Fedik - just added Global Config option to enable/disable device specific caching. 
Useful for website who don't differ the HTML output between mobile or non-mobile devices.

File: libraries/joomla/cache/cache.php [hunk 4]
```diff
@@ -748,7 +750,33 @@ public static function makeId()
...
+	 * @since 3.5
+	 */
+	public static function getDevicePrefix()
+	{
+		jimport('joomla.environment.browser');
+		$browser = JBrowser::getInstance();
+		$conf = JFactory::getConfig();
+
+		if (boolval(!$conf->get('deviceprefix_enable', '0')))
```
Comment (smz) [inline libraries/joomla/cache/cache.php]: I think `boolval()` should not be used as, AFAIK, it is available with PHP >= 5.5.0 only
Also the `!` negation operator should be applied to the result of the `boolval()` function...
A simpler soultion could be:
```
if ($conf->get('deviceprefix_enable', '0')))
{
	if ($browser->isMobile())
	{
		//return prefix for mobile
		return 'M-';
	}
}

// return empty string
return '';
```

Comment (hans2103) [inline libraries/joomla/cache/cache.php]: Last year I've seen a presentation by Rafael Dohms during #JAB14
He explained us not to use multiple indents. 
http://www.slideshare.net/rdohms/presentations

Therefor I will not add multiple if-statementes. But I will take another look into the booval.

**Comment by ggppdk** [comment]
I think this PR is quite useful for websites (or extensions),
that want to serve different page to mobile devices

not everything can be done in CSS

only issue that i see in this change, is that it will make cache system consume more space

**Comment by asika32764** [comment]
:+1:

**Comment by mbabker** [comment]
If we're really going to pursue this (I still don't see what the benefit here is personally), I'd suggest using [JApplicationWebClient](https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/application/web/client.php) (`$app->client`) instead of JBrowser as it is supposed to be a bit more 'intelligent' with platform detection.

File: administrator/components/com_config/model/form/application.xml [hunk 5]
```diff
@@ -25,6 +25,19 @@
 		</field>
 
 		<field
+			name="deviceprefix_enable"
```
Comment (mbabker) [inline administrator/components/com_config/model/form/application.xml]: Nitpicking at this point, but can we rename the param to `cache_deviceprefix`?  IMO that's a better short name to describe what is intended to be happening.  Or even `cache_platformprefix` as this isn't actually appending a device specific prefix, rather it's a platform level thing.

File: libraries/joomla/cache/cache.php [hunk 6]
```diff
@@ -713,6 +714,7 @@ public static function setWorkarounds($data, $options = array())
 	public static function makeId()
 	{
 		$app = JFactory::getApplication();
+		$prefix = JCache::getDevicePrefix();
```
Comment (mbabker) [inline libraries/joomla/cache/cache.php]: `$prefix = static::getDevicePrefix();`

Our code style rules state that when you're calling a static method in the same class (or a parent) that you should refer to it as self/static.

File: libraries/joomla/cache/cache.php [hunk 7]
```diff
@@ -748,7 +750,33 @@ public static function makeId()
 			$safeuriaddon->$key = $app->input->get($key, null, $value);
 		}
 
-		return md5(serialize($safeuriaddon));
+		return $prefix . md5(serialize($safeuriaddon));
```
Comment (mbabker) [inline libraries/joomla/cache/cache.php]: Is there any particular reason the prefix isn't included in the MD5 hash?

Comment (hans2103) [inline libraries/joomla/cache/cache.php]: when you exclude the prefix from the MD5 hash and set it before the MD5 hash you can easily filter out specific prefix

example:
MD5 hash for cache instance without prefix: asgsdgsduskngsdg9uw956w
cache instance for mobile : m-asgsdgsduskngsdg9uw956w
future pull request for tablet, handheld and fridge screen:
t-asgsdgsduskngsdg9uw956w
h-asgsdgsduskngsdg9uw956w
f-asgsdgsduskngsdg9uw956w

leaving prefix out of the MD5 hash let you easily filter.

**Comment by ggppdk** [comment]
about using JApplicationWebClient, the code should be (asking if i am correct)
```
$client = new JApplicationWebClient();
$isMobile = $client->__get('mobile');
```

About this topic someone can clame that is better done at the component as it should know better of what is happenning, e.g. see the following code,

- but the website integrator can be using custom template code, or using Joomla plugins, etc, that the component does not know about, so having this inside Joomla itself, makes sense


For doing it inside the component i have an example below
```
function display($cachable = null, $urlparams = false)
{
	...
	
	$jinput = JFactory::getApplication()->input;

	if (!empty($urlparams))
		$safeurlparams = & $urlparams;  // explicitely given
	else
	{
		$safeurlparams = array();  // Decide which URL params to add ...
		...
	}
	
	
	// If component is serving different pages to logged users, this will avoid
	// having users seeing same page after login/logout when conservative caching is used
	if ($userid = JFactory::getUser()->get('id'))
	{
		$jinput->set('__user_id__', $userid);
		$safeurlparams['__user_id__'] = 'STRING';
	}
	
	
	// If component is serving different pages for mobile devices, this will avoid
	// having users seeing the same page regardless of being on desktop or mobile
	$client = new JApplicationWebClient();
	if ($client->__get('mobile'))
	{
		$jinput->set('__mobile__', 'YES');
		$safeurlparams['__mobile__'] = 'STRING';
	}
	
	
	// ... decide if page is cacheable, according to logic known to the component
	$cachable = ...
	
	parent::display($cachable, $safeurlparams);
}
```

**Comment by mbabker** [comment]
You don't even need to instantiate it every time.

```php
$client = JFactory::getApplication()->client;
$isMobile = $client->mobile;
```

Typically you won't call methods with double underscores directly (like `__construct()`, `__get()`, etc.).  PHP calls those magic methods and handles using them when needed.  So `$client->mobile` is the same as your `$client->__get('mobile')`.

Otherwise I didn't really dig into your example, the client stuff just happened to show up immediately in the channel I'm seeing notifications from.

**Comment by ggppdk** [comment]
> You don't even need to instantiate it every time.

Thanks, i wanted to avoid making detection run twice
i now see the creation of the JApplicationWebClient object inside the JApplicationWeb class

- about __get you are again right ! thanks for the info !!

I missed this, here is some more info

For object properties that are to be very often accessed e.g. a big loop using 
- protected property  plus a  magic method  __get(),
will result in 5 times slower access than access using a custom "getter" function, because of the extra logic executed by PHP
(i speak of the access, of course the property calculation will only run once)

- thus a real getter:   $client->get('mobile'), would be much faster
- and both the above are even more slow that a public property,  which of course in our case we want to avoid

of course, someone getting this and using it an big loop should store it in a variable (if he/she is aware that the property is checked and returned by a magic function)

File: administrator/components/com_config/model/form/application.xml (hunk 5 above)
Comment (hans2103) [inline administrator/components/com_config/model/form/application.xml]: good suggestion. Will change in this file and other files depending on this param.

File: libraries/joomla/cache/cache.php (hunk 6 above)
Comment (hans2103) [inline libraries/joomla/cache/cache.php]: lessons learned. Thank you. Will change.

**Comment by hans2103** [comment]
Thank you for the feedback @mbabker and @ggppdk 
I've adjusted my commit.

**Comment by wojsmol** [comment]
@hans2103 If you want to add a parameter in the global configuration there will still be necessary changes in these 2 files
https://github.com/joomla/joomla-cms/blob/c41fcf023012741e230acac6acbe956f5c275c41/installation/model/configuration.php
https://github.com/joomla/joomla-cms/blob/c41fcf023012741e230acac6acbe956f5c275c41/installation/configuration.php-dist
Please correct me if I'm wrong.

**Comment by hans2103** [comment]
oh boy... I just came to a point that I need some serious help on coding. 
With the last commit I notice that I repeat myself with coding. 
I need some assistance with this pull request.

**Comment by hans2103** [comment]
found out that I had to set Headers vary:user-agent; to get it working. 
Live implementation with Redis caching can be seen by visiting https://www.byte.nl

**Comment by roland-d** [comment]
@mbabker Travis fails with an Application Instantiation Error, looking at where the test fails
```php
$this->object->store($data, $id, $group)
```

I think it may have something to do with this notice:

> PHP Warning:  PHP Startup: Unable to load dynamic library '/home/travis/.phpenv/versions/5.6.5/lib/php/extensions/no-debug-zts-20131226/apcu.so' - /home/travis/.phpenv/versions/5.6.5/lib/php/extensions/no-debug-zts-20131226/apcu.so: cannot open shared object file: No such file or directory in Unknown on line 0
Warning: PHP Startup: Unable to load dynamic library '/home/travis/.phpenv/versions/5.6.5/lib/php/extensions/no-debug-zts-20131226/apcu.so' - /home/travis/.phpenv/versions/5.6.5/lib/php/extensions/no-debug-zts-20131226/apcu.so: cannot open shared object file: No such file or directory in Unknown on line 0

As if no caching mechanism can be loaded. Is that the case?

**Comment by mbabker** [comment]
The PHP warnings have nothing to do with our `Application Instantiation Error` message.

**Comment by dgt41** [comment]
@roland-d can you restart travis here?

**Comment by brianteeman** [comment]
Restarted it for yoy

**Comment by hans2103** [comment]
next week (10 October) we have a PBF in the Netherlands. Roland and I will work on this pull request.

**Comment by hans2103** [comment]
Closing this PR in favor of #8055

**Pull request closed without merging**